### PR TITLE
Add finalizers to blueprint

### PR DESCRIPTION
**Pull request by shlomitk1**
This PR adds a finalizer to Blueprint. Deletion of the blueprint results in deleting the associated helm releases.
Fixes https://github.com/IBM/the-mesh-for-data/issues/63

### CODE REVIEW

File: manager/controllers/app/blueprint_controller.go [hunk 1]
```diff
@@ -51,41 +52,87 @@ func (r *BlueprintReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
 
-func (r *BlueprintReconciler) deleteChartResource(ctx context.Context, kubeNamespace string, releaseName string) (ctrl.Result, error) {
-	rel, err := r.Helmer.Status(kubeNamespace, releaseName)
-	if err == nil && rel != nil {
-		_, _ = r.Helmer.Uninstall(kubeNamespace, releaseName)
+// reconcileFinalizers reconciles finalizers for Blueprint
+func (r *BlueprintReconciler) reconcileFinalizers(blueprint *app.Blueprint) error {
+	// finalizer
+	finalizerName := r.Name + ".finalizer"
+	hasFinalizer := utils.ContainsFinalizer(blueprint.GetFinalizers(), finalizerName)
```
Comment (roee88) [inline manager/controllers/app/blueprint_controller.go:89]: I believe that we can use `ctrlutil.ContainsFinalizer` now (we use v0.6.2 of controller-runtime)

https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.2/pkg/controller/controllerutil/controllerutil.go#L290

File: manager/controllers/app/blueprint_controller.go [hunk 2]
```diff
@@ -51,41 +52,87 @@ func (r *BlueprintReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
-func (r *BlueprintReconciler) applyChartResource(ctx context.Context, log logr.Logger, ref string, vals map[string]interface{}, kubeNamespace string, releaseName string) (ctrl.Result, error) {
+func (r *BlueprintReconciler) deleteExternalResources(blueprint *app.Blueprint) error {
+	for _, step := range blueprint.Spec.Flow.Steps {
+		releaseName := getReleaseName(step)
+		rel, err := r.Helmer.Status(blueprint.Namespace, releaseName)
+		if err == nil && rel != nil {
+			_, _ = r.Helmer.Uninstall(blueprint.Namespace, releaseName)
+		}
+	}
+	return nil
```
Comment (roee88) [inline manager/controllers/app/blueprint_controller.go:132]: This seems like a best effort to delete, ignoring errors. Either:
1. Declare that this is a best effort and then you can avoid the call to Status (counting on the removal of the namespace to do the trick in most cases)
2. Track errors and if you encountered one then on the end of the loop don't return nil but rather an error (better IMHO)

File: manager/controllers/app/m4dapplication_controller.go [hunk 3]
```diff
@@ -108,7 +108,7 @@ func checkBlueprintStatus(applicationContext *app.M4DApplication, blueprint *app
 func (r *M4DApplicationReconciler) reconcileFinalizers(applicationContext *app.M4DApplication) error {
 	// finalizer (Blueprint)
 	finalizerName := r.Name + ".finalizer"
-	hasFinalizer := ContainsFinalizer(applicationContext, finalizerName)
+	hasFinalizer := utils.ContainsFinalizer(applicationContext.GetFinalizers(), finalizerName)
```
Comment (roee88) [inline manager/controllers/app/m4dapplication_controller.go:111]: same comment as above about `ContainsFinalizer`

File: manager/controllers/utils/utils.go [hunk 4]
```diff
@@ -216,3 +216,14 @@ func SupportsFlow(array []app.ModuleFlow, element app.ModuleFlow) bool {
...
+// ContainsFinalizer returns true if the given finalizer string appears in the list of finalizers of the object
+// Should be implemented in controller-runtime package - TO FIX
+func ContainsFinalizer(finalizers []string, finalizer string) bool {
+	for _, e := range finalizers {
+		if e == finalizer {
+			return true
+		}
+	}
+	return false
+}
```
Comment (roee88) [inline manager/controllers/utils/utils.go:229]: Same comment as above about `ContainsFinalizer`

File: manager/controllers/app/blueprint_controller.go [hunk 5]
```diff
@@ -51,41 +52,87 @@ func (r *BlueprintReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+			return err
+		}
+	}
+	return nil
+}
 
-	return ctrl.Result{}, nil
+func getReleaseName(step app.FlowStep) string {
+	// we add the "r" character at the beginning of the release name, since it must begin with an alphabetic character
+	return "r" + utils.Hash(step.Name, 20)
```
Comment (roee88) [inline manager/controllers/app/blueprint_controller.go:121]: unrelated to this PR but I think that we should start moving common code like `Hash` to appropriate folders in `pkg`

**Review by roee88 — COMMENTED** [review]
Looks good, minor comments.

File: manager/controllers/app/blueprint_controller.go (hunk 2 above)
Comment (shlomitk1) [reply manager/controllers/app/blueprint_controller.go]: Done.

File: manager/controllers/app/blueprint_controller.go (hunk 1 above)
Comment (shlomitk1) [reply manager/controllers/app/blueprint_controller.go]: Done.

File: manager/controllers/app/m4dapplication_controller.go (hunk 3 above)
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go]: Done.

File: manager/controllers/utils/utils.go (hunk 4 above)
Comment (shlomitk1) [reply manager/controllers/utils/utils.go]: Done.

File: manager/controllers/app/blueprint_controller.go [hunk 6]
```diff
@@ -51,41 +52,88 @@ func (r *BlueprintReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+func (r *BlueprintReconciler) deleteExternalResources(blueprint *app.Blueprint) error {
+	var err error
+	for _, step := range blueprint.Spec.Flow.Steps {
+		releaseName := getReleaseName(step)
+		if rel, errStatus := r.Helmer.Status(blueprint.Namespace, releaseName); errStatus != nil || rel == nil {
+			continue
+		}
+		_, err = r.Helmer.Uninstall(blueprint.Namespace, releaseName)
+	}
+	return err
```
Comment (roee88) [inline manager/controllers/app/blueprint_controller.go:133]: 1. Should we log errStatus for debugging?
2. This implementation simply returns what the last call to uninstall returns, which might be nil even if a previous uninstall call returned an error. You can either fix this directly (e.g., to only assign to the err var on the first actual error), or consider to group all returned errors (see implementation in https://github.com/helm/helm/blob/master/pkg/action/uninstall.go or the https://github.com/hashicorp/go-multierror library).

Comment (shlomitk1) [reply manager/controllers/app/blueprint_controller.go]: Fixed.

**Review by roee88 — APPROVED** [review]

**Pull request merged**